File: typing_trainer.go
// typing_trainer.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

var wordsMap = map[string]map[string][]string{
	"ru": {
		"easy":   {"кот", "нос", "рот", "дом", "лес", "море", "вода", "рука", "нога", "мир"},
		"medium": {"компьютер", "программа", "алгоритм", "интернет", "сервер", "сайт", "код", "данные"},
		"hard":   {"программирование", "искусственный интеллект", "коммуникация", "информация", "технология"},
	},
	"en": {
		"easy":   {"cat", "dog", "sun", "run", "jump", "tree", "book", "fish", "bird", "hand"},
		"medium": {"computer", "program", "algorithm", "internet", "server", "website", "code", "data"},
		"hard":   {"programming", "artificial intelligence", "communication", "information", "technology"},
	},
}

type Stats struct {
	BestWPM       float64 `json:"best_wpm"`
	Games         int     `json:"games"`
	TotalAccuracy float64 `json:"total_accuracy"`
}

type Trainer struct {
	lang   string
	level  string
	text   string
	start  time.Time
	end    time.Time
	input  string
	stats  Stats
	statsFile string
}

func NewTrainer(lang, level string) *Trainer {
	t := &Trainer{lang: lang, level: level}
	t.statsFile = filepath.Join(os.Getenv("HOME"), fmt.Sprintf(".typing_trainer_%s.json", lang))
	t.loadStats()
	t.generateText()
	return t
}

func (t *Trainer) loadStats() {
	data, err := os.ReadFile(t.statsFile)
	if err != nil {
		t.stats = Stats{}
		return
	}
	json.Unmarshal(data, &t.stats)
}

func (t *Trainer) saveStats() {
	data, _ := json.MarshalIndent(t.stats, "", "  ")
	os.WriteFile(t.statsFile, data, 0644)
}

func (t *Trainer) generateText() {
	words := wordsMap[t.lang][t.level]
	rand.Seed(time.Now().UnixNano())
	count := rand.Intn(6) + 5 // 5-10 слов
	if t.level == "hard" {
		count = rand.Intn(8) + 8 // 8-15
	}
	var builder strings.Builder
	for i := 0; i < count; i++ {
		if i > 0 {
			if t.level == "hard" && rand.Float32() < 0.3 {
				builder.WriteString(", ")
			} else {
				builder.WriteString(" ")
			}
		}
		word := words[rand.Intn(len(words))]
		builder.WriteString(word)
		if t.level == "hard" && (i == count-1 || rand.Float32() < 0.2) {
			builder.WriteString(". ")
		}
	}
	t.text = strings.TrimSpace(builder.String())
}

func (t *Trainer) displayText() {
	fmt.Println(colorize("\nНапечатайте следующий текст:", bold))
	fmt.Println(colorize(t.text, blue))
	fmt.Println("\nНажмите Enter после завершения ввода. Время начнётся с первой буквы.")
	fmt.Println("Для выхода нажмите Ctrl+C.\n")
}

func (t *Trainer) getUserInput() {
	reader := bufio.NewReader(os.Stdin)
	t.start = time.Now()
	input, _ := reader.ReadString('\n')
	t.input = strings.TrimRight(input, "\n")
	t.end = time.Now()
}

func (t *Trainer) calculateAndDisplay() {
	original := t.text
	typed := t.input
	total := len(original)
	typedLen := len(typed)
	correct := 0
	errors := 0
	for i := 0; i < total; i++ {
		if i < typedLen && original[i] == typed[i] {
			correct++
		} else {
			errors++
		}
	}
	if typedLen < total {
		errors += total - typedLen
	}
	elapsed := t.end.Sub(t.start).Seconds()
	minutes := elapsed / 60.0
	wpm := (float64(correct) / 5.0) / minutes
	if minutes == 0 {
		wpm = 0
	}
	accuracy := (float64(correct) / float64(total)) * 100.0

	// Цветной вывод
	fmt.Println(colorize("\nРезультат:", bold))
	for i, ch := range original {
		if i < typedLen && ch == rune(typed[i]) {
			fmt.Print(colorize(string(ch), green))
		} else {
			fmt.Print(colorize(string(ch), red))
		}
	}
	fmt.Println()

	fmt.Printf("%s: %.1f WPM\n", colorize("Скорость", yellow), wpm)
	fmt.Printf("Точность: %.1f%%\n", accuracy)
	fmt.Printf("Правильных символов: %d/%d\n", correct, total)
	fmt.Printf("Ошибок: %d\n", errors)
	fmt.Printf("Время: %.2f сек\n", elapsed)

	t.stats.Games++
	if wpm > t.stats.BestWPM {
		t.stats.BestWPM = wpm
		fmt.Println(colorize("🏆 Новый рекорд!", green))
	}
	t.stats.TotalAccuracy = (t.stats.TotalAccuracy*float64(t.stats.Games-1) + accuracy) / float64(t.stats.Games)
	t.saveStats()
	fmt.Printf("%s: %.1f WPM\n", colorize("Лучший результат", blue), t.stats.BestWPM)
}

func (t *Trainer) run() {
	fmt.Println(colorize("⌨️  Тренажёр слепой печати", bold))
	fmt.Printf("Язык: %s, Уровень: %s\n", t.lang, t.level)
	t.displayText()
	t.getUserInput()
	t.calculateAndDisplay()
}

func main() {
	lang := "ru"
	level := "medium"
	showStats := false
	resetStats := false
	args := os.Args[1:]
	for i := 0; i < len(args); i++ {
		arg := args[i]
		switch arg {
		case "ru", "en":
			lang = arg
		case "easy", "medium", "hard":
			level = arg
		case "-s", "--stats":
			showStats = true
		case "-r", "--reset":
			resetStats = true
		case "-h", "--help":
			fmt.Println("Usage: typing_trainer [ru|en] [easy|medium|hard] [-s] [-r]")
			return
		}
	}
	if resetStats {
		for _, l := range []string{"ru", "en"} {
			f := filepath.Join(os.Getenv("HOME"), fmt.Sprintf(".typing_trainer_%s.json", l))
			os.Remove(f)
		}
		fmt.Println("Статистика сброшена.")
		return
	}
	if showStats {
		for _, l := range []string{"ru", "en"} {
			f := filepath.Join(os.Getenv("HOME"), fmt.Sprintf(".typing_trainer_%s.json", l))
			data, err := os.ReadFile(f)
			if err != nil {
				fmt.Printf("Статистика для %s пуста.\n", l)
				continue
			}
			var stats Stats
			json.Unmarshal(data, &stats)
			fmt.Printf(colorize("📊 Статистика для %s:\n", bold), l)
			fmt.Printf("  Лучший WPM: %.1f\n", stats.BestWPM)
			fmt.Printf("  Сыграно игр: %d\n", stats.Games)
			fmt.Printf("  Средняя точность: %.1f%%\n", stats.TotalAccuracy)
		}
		return
	}
	trainer := NewTrainer(lang, level)
	trainer.run()
}

File: typing_trainer.cpp
// typing_trainer.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <random>
#include <chrono>
#include <fstream>
#include <sstream>
#include <cctype>
#include <filesystem>
#include <algorithm>

using namespace std;
using namespace std::chrono;
namespace fs = std::filesystem;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";
const string BOLD = "\033[1m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

string getHomeDir() {
    const char* home = getenv("HOME");
    if (!home) home = getenv("USERPROFILE");
    return string(home);
}

map<string, map<string, vector<string>>> WORDS = {
    {"ru", {
        {"easy", {"кот","нос","рот","дом","лес","море","вода","рука","нога","мир"}},
        {"medium", {"компьютер","программа","алгоритм","интернет","сервер","сайт","код","данные"}},
        {"hard", {"программирование","искусственный интеллект","коммуникация","информация","технология"}}
    }},
    {"en", {
        {"easy", {"cat","dog","sun","run","jump","tree","book","fish","bird","hand"}},
        {"medium", {"computer","program","algorithm","internet","server","website","code","data"}},
        {"hard", {"programming","artificial intelligence","communication","information","technology"}}
    }}
};

struct Stats {
    double best_wpm = 0;
    int games = 0;
    double total_accuracy = 0;
};

class Trainer {
public:
    string lang, level;
    string text;
    string userInput;
    high_resolution_clock::time_point start, end;
    Stats stats;
    string statsFile;

    Trainer(string l, string lvl) : lang(l), level(lvl) {
        statsFile = getHomeDir() + "/.typing_trainer_" + lang + ".json";
        loadStats();
        generateText();
    }

    void loadStats() {
        ifstream f(statsFile);
        if (!f) return;
        string content((istreambuf_iterator<char>(f)), istreambuf_iterator<char>());
        // Простой парсинг JSON (без библиотеки)
        auto extract = [&](const string& key) -> double {
            size_t pos = content.find("\"" + key + "\"");
            if (pos == string::npos) return 0;
            pos = content.find(":", pos) + 1;
            size_t end = content.find(",", pos);
            if (end == string::npos) end = content.find("}", pos);
            string val = content.substr(pos, end-pos);
            val.erase(remove_if(val.begin(), val.end(), ::isspace), val.end());
            if (val.front() == '"') val.erase(0,1);
            if (val.back() == '"') val.pop_back();
            return stod(val);
        };
        stats.best_wpm = extract("best_wpm");
        stats.games = (int)extract("games");
        stats.total_accuracy = extract("total_accuracy");
    }

    void saveStats() {
        ofstream f(statsFile);
        if (f) {
            f << "{\"best_wpm\":" << stats.best_wpm
              << ",\"games\":" << stats.games
              << ",\"total_accuracy\":" << stats.total_accuracy << "}";
        }
    }

    void generateText() {
        auto& words = WORDS[lang][level];
        random_device rd;
        mt19937 gen(rd());
        uniform_int_distribution<> dis(0, words.size()-1);
        int count = (level == "hard") ? rand() % 8 + 8 : rand() % 6 + 5;
        stringstream ss;
        for (int i=0; i<count; ++i) {
            if (i>0) {
                if (level == "hard" && (rand() % 10 < 3)) ss << ", ";
                else ss << " ";
            }
            ss << words[dis(gen)];
            if (level == "hard" && (i==count-1 || (rand() % 10 < 2))) {
                ss << ". ";
            }
        }
        text = ss.str();
        // Удаляем лишние пробелы в конце
        while (!text.empty() && text.back() == ' ') text.pop_back();
    }

    void displayText() {
        cout << colorize("\nНапечатайте следующий текст:", BOLD) << endl;
        cout << colorize(text, BLUE) << endl;
        cout << "\nНажмите Enter после завершения ввода. Время начнётся с первой буквы." << endl;
        cout << "Для выхода нажмите Ctrl+C.\n" << endl;
    }

    void getUserInput() {
        cin.ignore(); // очистка буфера
        start = high_resolution_clock::now();
        getline(cin, userInput);
        end = high_resolution_clock::now();
    }

    void calculateAndDisplay() {
        string original = text;
        string typed = userInput;
        int total = original.size();
        int typedLen = typed.size();
        int correct = 0;
        int errors = 0;
        for (int i=0; i<total; ++i) {
            if (i < typedLen && original[i] == typed[i]) {
                correct++;
            } else {
                errors++;
            }
        }
        if (typedLen < total) errors += total - typedLen;
        auto duration = duration_cast<milliseconds>(end - start).count();
        double elapsed = duration / 1000.0;
        double minutes = elapsed / 60.0;
        double wpm = (correct / 5.0) / minutes;
        double accuracy = (correct / (double)total) * 100.0;

        // Цветной вывод
        cout << colorize("\nРезультат:", BOLD) << endl;
        for (int i=0; i<total; ++i) {
            if (i < typedLen && original[i] == typed[i])
                cout << colorize(string(1, original[i]), GREEN);
            else
                cout << colorize(string(1, original[i]), RED);
        }
        cout << endl;

        cout << colorize("Скорость: " + to_string(wpm) + " WPM", YELLOW) << endl;
        cout << "Точность: " << accuracy << "%" << endl;
        cout << "Правильных символов: " << correct << "/" << total << endl;
        cout << "Ошибок: " << errors << endl;
        cout << "Время: " << elapsed << " сек" << endl;

        stats.games++;
        if (wpm > stats.best_wpm) {
            stats.best_wpm = wpm;
            cout << colorize("🏆 Новый рекорд!", GREEN) << endl;
        }
        stats.total_accuracy = (stats.total_accuracy * (stats.games-1) + accuracy) / stats.games;
        saveStats();
        cout << colorize("Лучший результат: " + to_string(stats.best_wpm) + " WPM", BLUE) << endl;
    }

    void run() {
        cout << colorize("⌨️  Тренажёр слепой печати", BOLD) << endl;
        cout << "Язык: " << lang << ", Уровень: " << level << endl;
        displayText();
        getUserInput();
        calculateAndDisplay();
    }
};

int main(int argc, char* argv[]) {
    string lang = "ru", level = "medium";
    bool showStats = false, resetStats = false;
    for (int i=1; i<argc; ++i) {
        string arg = argv[i];
        if (arg == "ru" || arg == "en") lang = arg;
        else if (arg == "easy" || arg == "medium" || arg == "hard") level = arg;
        else if (arg == "-s" || arg == "--stats") showStats = true;
        else if (arg == "-r" || arg == "--reset") resetStats = true;
        else if (arg == "-h" || arg == "--help") {
            cout << "Usage: typing_trainer [ru|en] [easy|medium|hard] [-s] [-r]" << endl;
            return 0;
        }
    }
    if (resetStats) {
        for (string l : {"ru", "en"}) {
            string f = getHomeDir() + "/.typing_trainer_" + l + ".json";
            if (fs::exists(f)) fs::remove(f);
        }
        cout << "Статистика сброшена." << endl;
        return 0;
    }
    if (showStats) {
        for (string l : {"ru", "en"}) {
            string f = getHomeDir() + "/.typing_trainer_" + l + ".json";
            ifstream file(f);
            if (file) {
                string content((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
                auto extract = [&](const string& key) -> double {
                    size_t pos = content.find("\"" + key + "\"");
                    if (pos == string::npos) return 0;
                    pos = content.find(":", pos) + 1;
                    size_t end = content.find(",", pos);
                    if (end == string::npos) end = content.find("}", pos);
                    string val = content.substr(pos, end-pos);
                    val.erase(remove_if(val.begin(), val.end(), ::isspace), val.end());
                    if (val.front() == '"') val.erase(0,1);
                    if (val.back() == '"') val.pop_back();
                    return stod(val);
                };
                double best = extract("best_wpm");
                int games = (int)extract("games");
                double acc = extract("total_accuracy");
                cout << colorize("📊 Статистика для " + l + ":", BOLD) << endl;
                cout << "  Лучший WPM: " << best << endl;
                cout << "  Сыграно игр: " << games << endl;
                cout << "  Средняя точность: " << acc << "%" << endl;
            } else {
                cout << "Статистика для " + l + " пуста." << endl;
            }
        }
        return 0;
    }
    Trainer trainer(lang, level);
    trainer.run();
    return 0;
}
